plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/ExprUtils.kt

```diff
@@ -5,13 +5,28 @@
 
 package org.jetbrains.kotlin.formver.purity
 
+import org.jetbrains.kotlin.KtSourceElement
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.formver.ErrorCollector
+import org.jetbrains.kotlin.formver.embeddings.expression.Assert
 import org.jetbrains.kotlin.formver.embeddings.expression.ExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.WithPosition
 
-fun ExpEmbedding.preorder(): Sequence<ExpEmbedding> = sequence {
-    yield(this@preorder)
-    yieldAll(this@preorder.children().flatMap { it.preorder() })
+class PositionedExpEmbedding(val embedding: ExpEmbedding, val source: KtSourceElement?)
+
+// Positioning information about a node is stored in the closest parent WithPosition node
+fun ExpEmbedding.preorder(currentSource: KtSourceElement? = null): Sequence<PositionedExpEmbedding> = sequence {
+    val nextSource: KtSourceElement? = when (this@preorder) {
+        is WithPosition -> this@preorder.source
+        else -> currentSource
+    }
+    yield(PositionedExpEmbedding(embedding = this@preorder, source = nextSource))
+
+    yieldAll(this@preorder.children().flatMap { it.preorder(currentSource = nextSource) })
 }
 
-fun ExpEmbedding.checkValidity(): Boolean =
-    preorder()
-        .all { it.isValid() }
\ No newline at end of file
+fun ExpEmbedding.checkValidity(declaration: FirSimpleFunction, errorCollector: ErrorCollector): Boolean =
+    preorder(declaration.source)
```

Comment:
If you only need the declaration for the source, why not pass the source directly?